rpc/grpc/api.go

```diff
@@ -37,3 +49,346 @@ func (bapi *broadcastAPI) BroadcastTx(ctx context.Context, req *RequestBroadcast
 		},
 	}, nil
 }
+
+type BlockAPI struct {
+	sync.Mutex
+	heightListeners      map[chan NewHeightEvent]struct{}
+	newBlockSubscription eventstypes.Subscription
+}
+
+func NewBlockAPI() *BlockAPI {
+	return &BlockAPI{
+		// TODO(rach-id) make 1000 configurable if there is a need for it
+		heightListeners: make(map[chan NewHeightEvent]struct{}, 1000),
+	}
+}
+
+func (blockAPI *BlockAPI) StartNewBlockEventListener(ctx context.Context) {
+	env := core.GetEnvironment()
+	if blockAPI.newBlockSubscription == nil {
+		var err error
+		blockAPI.newBlockSubscription, err = env.EventBus.Subscribe(
+			ctx,
+			fmt.Sprintf("new-block-grpc-subscription-%s", rand.Str(6)),
+			eventstypes.EventQueryNewBlock,
+			500,
+		)
+		if err != nil {
+			env.Logger.Error("Failed to subscribe to new blocks", "err", err)
+			return
+		}
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-blockAPI.newBlockSubscription.Cancelled():
+			env.Logger.Error("cancelled grpc subscription. retrying")
+			ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+			if err != nil {
+				blockAPI.closeAllListeners()
+				return
+			}
+			if !ok {
+				// this will happen when the context is done. we can stop here
+				return
+			}
+		case event, ok := <-blockAPI.newBlockSubscription.Out():
+			if !ok {
+				env.Logger.Error("new blocks subscription closed. re-subscribing")
+				ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+				if err != nil {
+					blockAPI.closeAllListeners()
+					return
+				}
+				if !ok {
+					// this will happen when the context is done. we can stop here
+					return
+				}
+				continue
+			}
+			newBlockEvent, ok := event.Events()[eventstypes.EventTypeKey]
+			if !ok || len(newBlockEvent) == 0 || newBlockEvent[0] != eventstypes.EventNewBlock {
+				continue
+			}
+			data, ok := event.Data().(eventstypes.EventDataNewBlock)
+			if !ok {
+				env.Logger.Debug("couldn't cast event data to new block")
+				continue
+			}
+			blockAPI.broadcastToListeners(ctx, data.Block.Height, data.Block.Hash())
+		}
+	}
+}
+
+func (blockAPI *BlockAPI) retryNewBlocksSubscription(ctx context.Context) (bool, error) {
+	env := core.GetEnvironment()
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	blockAPI.Lock()
+	defer blockAPI.Unlock()
+	for i := 1; i < 6; i++ {
```

Comment:
[question] out of curiosity why retry 5 times? Should we extract a constant for `RETRY_ATTEMPTS = 5`

```suggestion
	for i := 0; i < RETRY_ATTEMPTS; i++ {
```

Reply:
yes, we could. I can extract it if you want

Reply:
IMO an extracted comment is easier to read but this comment is an optional nit and not blocking.

Reply:
a8f8a21b720122801582dfa47e278db1ec372c0c